---
title: "Maize UniformMu mutants characterization"
output: rmarkdown::github_document
# author:
# - name: Peng Zhou
#  affiliation: University of Minnesota
# - name: Nathan Springer
#  affiliation: University of Minnesota
# abstract: "This document provides an introduction to R Markdown, argues for its..."
# keywords: "pandoc, r markdown, knitr"
date: "`r format(Sys.time(), '%B %d, %Y')`"
geometry: margin=1in
fontfamily: mathpazo
fontsize: 11pt
# spacing: double
# bibliography: ~/Dropbox/master.bib
# biblio-style: apsr
header-includes:
- \usepackage{caption}
- \usepackage{booktabs}
- \usepackage{longtable}
- \usepackage{array}
- \usepackage{multirow}
- \usepackage[table]{xcolor}
- \usepackage{wrapfig}
- \usepackage{float}
- \usepackage{colortbl}
- \usepackage{pdflscape}
- \usepackage{tabu}
- \usepackage{threeparttable}
- \usepackage{threeparttablex}
- \usepackage[normalem]{ulem}
- \usepackage{makecell}
- \captionsetup[figure]{labelformat=empty}
- \captionsetup[table]{labelformat=empty}
- \AtBeginDocument{\let\maketitle\relax}
---
\pagenumbering{gobble}

Raw GFF3 file mapping to B73 AGPv3 were obtained from maizeGDB
  - 79,407 insertion sites and related stocks information

Insertion sites were lifted over to AGPv4
  - 77,085 sites

The overlap of these sites with AGPv4 exons were checked:
  - 36,648 sites overlap w. at least one exon(s)
  - 38,778 sites do not overlap w. an exon (i.e., either in introns, UTRs or intergenic spaces)

A total of 14,831 genes have at least one exonic insertion sites, 9,046 have at least two exonic insertions.

```{r echo = F, message = F, warning = F}
require(tidyverse)
require(knitr)
require(kableExtra)
options(knitr.kable.NA = '')
dirw = '~/projects/genomes/data/uniformmu' 
fi = file.path(dirw, '15.uniformmu.exon.rda')
x = load(fi)
```

Table 1.  Summary of genes with at least one exonic UniformMu insertion sites.
```{r echo = F, message = F, warning = F}
tt = ti %>% count(n_mu)
kable(tt, format = 'markdown', booktabs = T, 
    align = 'r',
    col.names = c('# exonic insertions', '# genes'))
```

Structural integrity of these 14,831 B73 genes in the W22 genome were checked based on variants called by whole genome comparison between B73 and W22. 10,236 genes have either no, low (such as synonymous change) or moderate (such as mis-sense) levels of changes in W22 compared to B73.

Table 2. Summary of gene changes between B73 and W22.
```{r echo = F, message = F, warning = F}
tt = ti %>% count(impact)
kable(tt, format = 'markdown', booktabs = T, 
    align = 'r',
    col.names = c('impact', '# genes'))
```

The expression of these genes in 23 tissues of B73 and Mo17 were checked, with genes showing differential expression between B73 and Mo17 in at least 5 tissues prioritized.

Table 3. Summary of genes showing DE in different number of tissues between B73 and Mo17.
```{r echo = F, message = F, warning = F}
tt = ti %>% count(de_BM)
kable(tt, format = 'markdown', booktabs = T, 
    align = 'r',
    col.names = c('# tissues DE btw. B73 and Mo17', '# genes'))
```

Finally, 826 of these genes are transcription factor based on [PlantTFDB](http://planttfdb.cbi.pku.edu.cn/index.php?sp=Zma).

[This table](15.uniformmu.exon.tsv) lists all 14,831 genes with at least one exonic UniformMu insertion sites, with columns:
 - `gid`: AGPv4 gene ID
 - `n_mu`: number exonic UniformMu insertion sites
 - `mid`, `sids`: mutant ID and stock ID
 - `tf`: whether this gene is TF
 - `ttype`: type of gene, ncRNA or protein-coding mRNA
 - `impact`, `eff`: impact and effect of the change from B73 to W22
 - `de_BM`: number of tissues showing DE between B73 and Mo17

A total of 339 TFs have at least 2 exonic insertions and have gene structure conserved in the W22 genome.

Table 4.  Summary of the 339 TFs by number DE tissues between B73 and Mo17
```{r echo = F, message = F, warning = F}
low_impacts = c("no_change","low",'modifier','moderate')
tt = ti %>% filter(tf, n_mu != '1', impact %in% low_impacts) %>% count(de_BM)
kable(tt, format = 'markdown', booktabs = T, 
    align = 'r',
    col.names = c('# tissues DE btw. B73 and Mo17', '# genes'))
```


