---
title: "Maize UniformMu mutants characterization"
output: rmarkdown::github_document
# author:
# - name: Peng Zhou
#  affiliation: University of Minnesota
# - name: Nathan Springer
#  affiliation: University of Minnesota
# abstract: "This document provides an introduction to R Markdown, argues for its..."
# keywords: "pandoc, r markdown, knitr"
date: "`r format(Sys.time(), '%B %d, %Y')`"
geometry: margin=1in
fontfamily: mathpazo
fontsize: 11pt
# spacing: double
# bibliography: ~/Dropbox/master.bib
# biblio-style: apsr
header-includes:
- \usepackage{caption}
- \usepackage{booktabs}
- \usepackage{longtable}
- \usepackage{array}
- \usepackage{multirow}
- \usepackage[table]{xcolor}
- \usepackage{wrapfig}
- \usepackage{float}
- \usepackage{colortbl}
- \usepackage{pdflscape}
- \usepackage{tabu}
- \usepackage{threeparttable}
- \usepackage{threeparttablex}
- \usepackage[normalem]{ulem}
- \usepackage{makecell}
- \captionsetup[figure]{labelformat=empty}
- \captionsetup[table]{labelformat=empty}
- \AtBeginDocument{\let\maketitle\relax}
---
\pagenumbering{gobble}

Raw GFF3 file mapping to B73 AGPv3 were obtained from maizeGDB
  - 79,407 insertion sites and related stocks information

Insertion sites were lifted over to AGPv4
  - 77,085 sites

The overlap of these sites with AGPv4 exons were checked:
  - 36,648 sites overlap w. at least one exon(s)
  - 38,778 sites do not overlap w. an exon (i.e., either in introns, UTRs or intergenic spaces)

A total of 14,831 genes have at least one exonic insertion sites, 9,046 have at least two exonic insertions.

- Table 1.  Summary of genes with at least one exonic UniformMu insertion sites.

```{r message=FALSE, warning=FALSE}
require(tidyverse)
require(knitr)
require(kableExtra)
options(knitr.kable.NA = '')
dirw = '~/projects/genomes/data/uniformmu' 
fi = file.path(dirw, '/15.uniformmu.exon.tsv')
ti = read_tsv(fi)
tt = ti %>% count(n_mu)
kable(tt, format = 'markdown', booktabs = T, 
    align = 'r',
    col.names = c('# exonic insertions', '# genes'),
    caption = "Table 1.  Alignment statistics.") #%>%
  #column_spec(1, bold=T) %>%#, width_min='7cm')
```

